File: src/Tracker.cpp
#include <lacrosse-bot/tracker/Tracker.hpp>


namespace nurc
{

using namespace cv;
using namespace std;

Tracker::Tracker()
{
	video_capture_ = VideoCapture(0);
	if(!video_capture_.isOpened()) {
		image_ = Mat( video_capture_.get(CV_CAP_PROP_FRAME_HEIGHT), 
												 video_capture_.get(CV_CAP_PROP_FRAME_WIDTH),
												 CV_8UC3 );
	}
	else {
		image_ = Mat( 600, 800, CV_8UC3 );
	}

	previous_tick_count_ = current_tick_count_ = getTickCount();
	total_tick_count_ = 0;
	
}

bool Tracker::setVideoCapture(unsigned int capture)
{
	video_capture_.open(capture);
	return video_capture_.isOpened();
}

bool Tracker::setVideoCapture(const char* capture)
{
	video_capture_.open(capture);
	if(video_capture_.isOpened()) {
		image_.create( video_capture_.get(CV_CAP_PROP_FRAME_HEIGHT), 
												  video_capture_.get(CV_CAP_PROP_FRAME_WIDTH),
												  CV_8UC3 );
	}
	else {
		image_.create( 600, 800, CV_8UC3 );
	}
	return video_capture_.isOpened();
}

} // namespace nurc

File: scripts/Tracker.py
#/usr/bin/env python

import cv2
import numpy as np

class Tracker(object):
  
  def __init__(self):
    self.video_capture = cv2.VideoCapture(0)
    if self.video_capture.isOpened():
      self.valid, self.image = self.video_capture.read()
    else:
      self.image = np.empty( (600, 800, 3), float )
      self.valid = False
    self.previous_tick_count = self.current_tick_count = cv2.getTickCount()
    self.total_tick_count = 0
  
  def __del__(self):
    # Since OpenCV uses a binding interface should perform basic memory management
    self.video_capture.release()
  
  def setVideoCapture(self, capture):
    self.video_capture.open(capture)
    if isinstance(capture, str):
      self.image = np.empty( self.video_capture.get(cv2.cv.CV_CAP_PROP_FRAME_HEIGHT), self.video_capture.get(cv2.cv.CV_CAP_PROP_FRAME_WIDTH), 3 );
    else:
      self.image = np.empty( (600, 800, 3), float )
    return self.video_capture.isOpened()
  
  def captureImageFrame(self):
    if self.video_capture.isOpened():
      self.valid, self.image = self.video_capture.read()
      self.prev_tick_count = self.current_tick_count
      self.current_tick_count = cv2.getTickCount()
      self.total_tick_count += self.current_tick_count - self.prev_tick_count
  
  def getRuntime(self):
    return self.total_tick_count/cv2.getTickFrequency()
  


if __name__ == '__main__':
  # Unit tests
  tracker_test = Tracker()
  while cv2.waitKey(1) == -1 and tracker_test.valid:
    cv2.imshow("Tracker", tracker_test.image)
    tracker_test.captureImageFrame()
  cv2.destroyAllWindows()
  
